### DISCUSSION TITLE
Add a new mode when tooltip is visible.

### DISCUSSION THREAD

**Discussion by Ergus** [company-mode/company-mode#1299] [Ideas]
Hi:

I think I have some "decent proposal" after trying the new `company-complete-common-or-show-delayed-tooltip`.

1) I have a config where I use M-TAB to show and hide company-tooltip. The idea behind is to have a sort of toggle effect. So, when using the new function and config I get a very uncomfortable behavior because I need to press M-TAB 2 or 3 times to get the tooltip because (company-abort is called in between).
2) Another issue with this is that when only preview is shown I cannot use C-n or C-p to move normally (because it is also bound in my `company-active-map`. (this is specially useful with the `company-preview-common-frontend`.

I added temporally some extra `if` here and there to a wrapper function to solve this but I think that a more "elegant" solution may be to add an extra mode when tooltip becomes visible ​something like company-tooltip-mode; that will add it's own map.

This solves the two issues because:
1) I could bind company-abort to M-TAB to the tooltip-map and keep the new `company-complete-common-or-show-delayed-tooltip` in the company-active-map.
2) We can bind C-n and C-p only in the new map so when only the preview is active the normal commands may not be affected.

As a good side effect here is that some commands that rely on `company-tooltip-visible-p` can simply be substituted with binding the same keybind to different commands in the different keymaps.

WDYT?

**Comment by yugaego** [comment]
Hi @Ergus. First of all, let me link the discussion that [preceded the addition of this function upstream](https://github.com/company-mode/company-mode/discussions/1214). It might provide more context/ideas to the topic discussed.

> Another issue with this is that when only preview is shown I cannot use C-n or C-p to move normally (because it is also bound in my company-active-map. 

For this  @dgutov had [suggested a workaround](https://github.com/company-mode/company-mode/discussions/1214#discussioncomment-1387585) I personally enjoy using. It is mentioned as well at the [Wiki Tips & Tricks](https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#enable-bindings-conditionally-for-a-shown-tooltip).  

> I could bind company-abort to M-TAB

Would it work if you'd bind it conditionally (via filter) too?

**Reply by Ergus** [reply]
Hi @yugaego:

> For this @dgutov had [suggested a workaround](https://github.com/company-mode/company-mode/discussions/1214#discussioncomment-1387585) I personally enjoy using. It is mentioned as well at the [Wiki Tips & Tricks](https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#enable-bindings-conditionally-for-a-shown-tooltip).
> 

I already have that for some of my commands, but in general I find myself remapping and wrapping almost all the commands in the `company-active-map`. This is complicated and undesirable.

Conceptually the issue is that showing `preview` and showing `pseudo-tooltip` are two very different states right? And we expect to interact with them in different modes: during preview the interaction must be minimally affected, while with the tooltip some bindings must be taken to select candidates and so on. This means tht when tooltip is visible most of the commands in `company-active-map` makes sense, but when only the preview is shown we only need `TAB` or `RET` to `company-complete` or `complete-common`. 

For example after the completion of the common part with `<tab>` the "preview-mode" shouldn't be enabled anymore, so a new `<tab>` should do whatever is bind in the global map; but actually as the timer is running it will execute whatever is in the `company-active-map`. So we end up with different combinations here according to the user preference each one will requiere a new function. (complete then indent. complete then show tooltip, complete and show tooltip then select next, only complete, complete common and abort, and so on)

> > I could bind company-abort to M-TAB
> 
> Would it work if you'd bind it conditionally (via filter) too?

No it doesn't because I need a more complex approach to show the tooltip when the `preview` is visible but the tooltip is not (basically put `company-tooltip-idle-delay` to  zero and then call `company-complete`). With the filter only the tooltip is not shown.

**Reply by yugaego** [reply]
I've been thinking about the answer for quite some time, and I still don't have a definite one.
 
I've got a feeling  that the initially suggested idea "to add an extra mode when tooltip becomes visible ​something like company-tooltip-mode; that will add it's own map" might be quite complex to implement (without introducing breaking changes). But I may be wrong. And I wonder, if such a change could resolve some other known issues. (Nothing comes to my mind at the moment.)

@dgutov , what do you think?  I'm sure you might have an idea on how to better approach this. Thanks.

@Ergus, by a chance, did you try implementing this idea (as a proof of the concept)
 and making sure it resolves the issues you're experiencing?
Posting the code experiments you've been mentioning might also simplify communication/considerations. 

> Conceptually the issue is that showing preview and showing pseudo-tooltip are
two very different states right?

I think we can say they're different at some level but not totally. They still
both are frontends, playing a role of displaying completion candidates.
Also, introducing any frontends-related changes, we'd have to consider all the Company frontends,
not only the ones mentioned, right?
   
> but when only the preview is shown we only need TAB or RET to company-complete
or complete-common.

This depends, a user may (I do, f.i.) still utilize other -- default -- bindings/commands when
only the preview is shown.

> For example after the completion of the common part with <tab> the
"preview-mode" shouldn't be enabled anymore, so a new <tab> should do whatever
is bind in the global map;

Isn't it a matter of a user need/taste/usage scenario?

> So we end up with different combinations here according to the user preference
each one will requiere a new function.

That's right. The use-cases/habits/needs differ a lot. It *seems to me*, that
each library/package/framework is an attempt to get to some "middle ground" that
would suit more or less majority of the use cases. So if there's a way to
improve Company in a sense of being more flexible and convenient to use for some
known use-cases, without breaking existing functionality, then it's probably a
way to go. (Given there's someone ready to implement those changes
and the price of the following support is not too high.
Again, that's my current humble opinion only.)

**Reply by dgutov** [reply]
I don't have an easy fix that would be consistent and user-friendly at the same time.

The problem (or what could be considered a problem) is that the UI extensibility uses the notion of "frontends" which ot1h visualize some underlying state of the completion process, but ot2h are completely invisible to the "main" code. There are some minor references between them (mainly to ensure that overlay-based rendering keeps working correctly), but otherwise Company's commands don't know whether a certain frontend is active at a given time. And the user can easily swap frontends and/or disable some of the built-in ones.

Another part of this (and a consequence of the above) is that the tooltip-with-delay feature is implemented as a frontend as well. And the caller code *really* doesn't know whether a popup should be visible at a given time: the delay logic is hidden inside the wrapper frontend.

And okay, if the only popup implementation frontends used was the `company-pseudo-tooltip-frontend`, we could add a minor mode which would be toggled by `company-pseudo-tooltip-unhide` and `company-pseudo-tooltip-hide` (*). But there are at least two more popup frontends (third-party ones) which some users prefer over the built-in one. And we similarly can't control their behavior except by coming there and asking to please do this one more thing... which kind of smells as a design: you create this really simple protocol, but then add these peculiar requirements on top of it. Which might still break from time to time, as the core code is unable to enforce them.

Or we go a longer route and change the set of events the frontends are supposed to handle. Then those could include such events as show-popup, show-preview, hide-popup, hide-preview. It will take a fair amount of work for everyone involved and a significant loss of flexibility as well. And, at least offhand, I'm not sure how the tooltip-with-delay would be implemented in that model.

But if we're just talking personal customization, @Ergus you should be able to do what I described above at (*): define your own buffer-local minor mode with a keymap, and add `:after` advices to `company-pseudo-tooltip-unhide` and `company-pseudo-tooltip-hide` to toggle it. If that is the frontend you're using, that is.

**Reply by Ergus** [reply]
Hi @dgutov:

On one hand the solution you propose is not terrible. I mean, a frontend may need to update itseld if the "framework" api changes, that's how it works. Maybe a package maintainer requires some more "complex" changes but at least it won't enforce the user to do it.
I have been working in tex recently and sadly I found that a slow backend with the config I show bellow is almost impossible to use. It feels like emacs hangs constantly when pressing some letter, so I need to wait ot press `C-g` constantly. The worst is that the user does not see anything because the preview is not shown (there is not common part) but the timer for the tooltip is running..

On the other hand, only frontends with extra delays are affected by this. So if the users/developers add just normal frontends without extra delays there is not difference with preview and tooltip state as they will visible be together. 

BTW: what are the other frontends you refer?

**Reply by dgutov** [reply]
> I mean, a frontend may need to update itseld if the "framework" api changes, that's how it works.

It might seem like a fairly weird API, though. If we're talking about having all popup frontends toggle a particular minor mode on show/hide.

> I have been working in tex recently and sadly I found that a slow backend with the config I show bellow is almost impossible to use.

Huh, so you mean that they try to press `C-n` and `C-p`, and see nothing happening, because the popup is not there? Perhaps the answer is to use a different preview frontend: just `company-preview-frontend` instead of `company-preview-if-just-one-frontend` or `company-preview-common-frontend`.

Or alternatively, the -with-delay frontend could be accompanied by a minor mode which would set up the alternative set of bindings. One that checks that the popup is visible, for example. So those `:filter` instructions would be all pre-set without requiring that effort from the user.

Would that solve the problems with your "slow tex backend", though? It would only take care about the conflicting bindings. If the problem is no visible indication of completion being in-progress, perhaps you just need to use different frontends. Like the default set.

> BTW: what are the other frontends you refer?

`company-posframe` and `company-box`.

**Reply by Ergus** [reply]
> Huh, so you mean that they try to press `C-n` and `C-p`, and see nothing happening, because the popup is not there? Perhaps the answer is to use a different preview frontend: just `company-preview-frontend` instead of `company-preview-if-just-one-frontend` or `company-preview-common-frontend`.
> 

In this case then all the `-with-delay` frontend should be considered broken and probably removed... because this problem is associated with the delay between `company-mode` and it's map becomes active and when the user sees something. In case the user don't use any preview or uses the common preview and has completed commons the issue is exposed as a `hang` for any binding `captured` by the `company-active-map`.

In general activating different bindings in every state (mode) may solve the issue, and putting the right parents to those maps, the changes doesn't look so complicated.

**Comment by Ergus** [comment]
BTW: 
(Maybe a bit offtopic)

In the current implementation what's the `right` condition to detect if the preview is not visible... I mean, after a first `<tab>` and completed-common, the preview overlay is not visible with the `company-preview-common-frontend` but the `company-active-map` is still active, so commands like `<tab>` need some kind of filter because this can be a bit confusing for the user unaware of what's going on.

I tried to use some condition with `company-preview-overlay` but somehow it seems to be always `nil` even when the preview is visible.  So I cannot use a `visible-p` or compare if the start equals the end.

Now the only condition I use is using `(buffer-chars-modified-tick)` which seems a bit dirty.

**Reply by yugaego** [reply]
> so commands like <tab> need some kind of filter because this can be a bit confusing for the user unaware of what's going on.

Maybe. Though consider  `company-complete-common-or-show-delayed-tooltip` has to be specifically bound/called, so some kind of awareness could be expected, I think.

> I tried to use some condition with company-preview-overlay but somehow it seems to be always nil even when the preview is visible. 

This sounds odd.

> Now the only condition I use is using (buffer-chars-modified-tick) which seems a bit dirty.

Do you mean at all, or with that particular case you're using it for?

**Reply by dgutov** [reply]
`company-preview-overlay` being non-nil should indeed be an indicator for the preview being visible.

Just keep in mind that it's a buffer-local variable: if your code is running in a different buffer, the variable can be nil there but non-nil in the original buffer.

**Reply by Ergus** [reply]
> `company-preview-overlay` being non-nil should indeed be an indicator for the preview being visible.
> 
> Just keep in mind that it's a buffer-local variable: if your code is running in a different buffer, the variable can be nil there but non-nil in the original buffer.

I know that, but it does not work. I just put a print (message) in some of the wrappers or filter functions to print the value of the overlay (preview or tooltip) and even when they are visible I got nil....)

**Reply by dgutov** [reply]
There are only two functions that do `(setq company-preview-overlay ...`.

The first is `company-preview-show-at-point`, and the second is `company-preview-hide`.

What have you tried exactly?

**Reply by Ergus** [reply]
Maybe it has to do with the fact that `company-preview-show-at-point` actually puts another overlay `ov` or with the `(defvar company-preview-overlay)` in company-pseudo-tooltip-unless-just-one-frontend-with-delay... You can just put a message and see the values are `nil` or not visible even with the preview active when using the `with-delay` frontend.

**Reply by dgutov** [reply]
`company-pseudo-tooltip-unless-just-one-frontend-with-delay` just checks that that overlay exists, and shows the popup right away if it doesn't. It doesn't modify that var.

If you can write down your scenario exactly step by step (which functions to modify and what to do), I can check it for you.

**Comment by Ergus** [comment]
Hi @yugaego:

Just in case you want to know what I have:

```
(setq-default company-minimum-prefix-length 1
		company-tooltip-minimum 5
		company-selection-wrap-around nil
		company-show-quick-access 'left       ;; How hints
		company-tooltip-align-annotations t
		company-format-margin-function #'company-detect-icons-margin
		company-require-match nil
		company-dabbrev-code-other-buffers nil

		company-idle-delay 0
		company-tooltip-idle-delay 100
		company-frontends '(company-pseudo-tooltip-unless-just-one-frontend-with-delay
				                  company-preview-common-frontend )

		company-backends '(company-capf))

  (defun my/filter-with-ptwd (command)
    "Return a COMMAND if a tooltip is shown; otherwise return nil."
    (if (company-tooltip-visible-p)
	command
      (lambda ()
	(interactive)
	(company-abort)
	(company--unread-this-command-keys))))

  (defun my/reverse-filter-with-ptwd (command)
    "Return a COMMAND if a tooltip is shown; otherwise return nil."
    (if (company-tooltip-visible-p)
	#'company-abort
      command))

  (defun my/company-complete ()
    (interactive)
    (let ((company-tooltip-idle-delay 0.0))
      (company-complete)
      (and company-candidates
           (company-call-frontends 'post-command))))

  (defun my/company-complete-common ()
    (interactive)
    (let ((old-tick (buffer-chars-modified-tick)))
      (call-interactively 'company-complete-common)
      (when (eq old-tick (buffer-chars-modified-tick))
	(if (company-tooltip-visible-p)
	    (company-select-next-or-abort)
          (company-abort)
	  (company--unread-this-command-keys)))))

  (keymap-set company-mode-map "M-RET" #'my/company-complete)
  (keymap-set company-mode-map "M-/" #'company-other-backend)

  (keymap-set company-active-map "<remap> <company-select-next-or-abort>"
	      #'(menu-item "" company-select-next-or-abort :filter my/filter-with-ptwd))
  (keymap-set company-active-map "<remap> <company-select-previous-or-abort>"
	      #'(menu-item "" company-select-previous-or-abort :filter my/filter-with-ptwd))
  (keymap-set company-active-map "<remap> <xref-find-definitions>"
	      #'(menu-item "" company-show-location :filter my/filter-with-ptwd))

  (keymap-set company-active-map "M-/" #'company-other-backend)

  (keymap-set company-active-map "M-RET"
	      #'(menu-item "" my/company-complete :filter my/reverse-filter-with-ptwd))

  (keymap-set company-active-map "<remap> <company-complete-common>" #'my/company-complete-common)
  (keymap-set company-active-map "<remap> <completion-at-point>" #'company-select-previous-or-abort)
```

**Reply by dgutov** [reply]
The above doesn't look too bad. We're likely to add more `:filter` conditions to the default bindings as well, to minimize the use of `company--unread-this-command-keys`.

BTW, you don't need to use it in `my/filter-with-ptwd`, do you?

Return nil, maybe even call `company-abort`, and Emacs will dispatch the same event to the next binding.

**Reply by Ergus** [reply]
> The above doesn't look too bad.

Here we disagree. Requiring this from the user is terrible. But also it does not work well when some backend is slow (ex: dabbrev + tex).

> We're likely to add more `:filter` conditions to the default bindings as well, to minimize the use of `company--unread-this-command-keys`.
> 
Suggestions?

Don't you think that with two different maps it may be simpler, cleaner and potentially more `efficient`? I mean, with an auxiliar mode or using `set-transient-map` for preview and a minor mode for tooltip? 

Basically all this is a consequence of the `*-with-delay` frontend... For the other frontends there is not problem because tooltip and preview are together. So maybe only the delayer frontends need to be affected?

> BTW, you don't need to use it in `my/filter-with-ptwd`, do you?
> Return nil, maybe even call `company-abort`, and Emacs will dispatch the same event to the next binding.

I will try, basically I took this code from *-or-abort commands.

**Reply by Ergus** [reply]
I just tried removing `company--unread-this-command-keys` from `my/filter-with-ptwd`. In general it required pressing (for example) `<down>` two times. One to abort company and the other to move down. This may be fine if the preview is visible, because we could assume that the first `<down>` hides the preview. But sometimes the preview is active, but not visible (like when `company-preview-common-frontend` but `complete-common` was already called. In that case the user needs an extra `<down>` without any visible reason.

**Reply by dgutov** [reply]
> Basically all this is a consequence of the *-with-delay frontend... For the other frontends there is not problem because tooltip and preview are together. So maybe only the delayer frontends need to be affected?

The `-with-delay` frontend is special indeed. Like I said above, perhaps we could automate some additional setup for the user, to get the most benefit from it. For now the code is on the wiki: https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#enable-bindings-conditionally-for-a-shown-tooltip

**Reply by dgutov** [reply]
> I just tried removing company--unread-this-command-keys from my/filter-with-ptwd. In general it required pressing (for example) <down> two times. One to abort company and the other to move down.

I'll look into this, thanks. This shouldn't be the case.

**Reply by dgutov** [reply]
@Ergus BTW, could you try replacing `my/filter-with-ptwd` with `my-company-keymap--filter-by-tooltip` from that wiki page?

**Reply by Ergus** [reply]
> @Ergus BTW, could you try replacing `my/filter-with-ptwd` with `my-company-keymap--filter-by-tooltip` from that wiki page?

That used to work partially, but I wanted a slightly different behavior. Basically I wanted to disable and abort company with arrows or `M-RET`, but providing  a toggle like behavior. When preview is active (even when not visible after a `complete-common`) and the tooltip is not visible yet, the filter `my-company-keymap--filter-by-tooltip` just ignores the arrows. 

As described before, after a `complete-common` the preview may be active, but not visible if there is not common part to all the candidates. If the user presses `<down>` if just feels that emacs is not responding and nothing happens until he presses `C-g` or the tooltip comes out.

**Reply by dgutov** [reply]
I don't understand, sorry.

There are basically two cases: when you want `<down>` to switch between the completions, and when you want it to move point to the next line. There is some condition logic to choose between the two (and we can discuss that), but as soon as you have decided to do the latter, you shouldn't have to call `company-abort` and `company--unread-this-command-keys`... right? Just having `next-line` invoked should suffice, because Company will see the new position right after and it will disable itself.
